### PR TITLE
Complete refactor of ruby code

### PR DESCRIPTION
**Pull request by dkniffin**
### Why?

- It was in need of a refactor.

### What changed?

- Completely refactored the ruby code for tape. It now uses thor for the command line interface, and it's much more organized.

### CODE REVIEW

**Comment by dkniffin** [comment]
Nobody merge this yet. I'm working on adding some tests for it.

File: lib/taperole/commands/ansible.rb [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+                   type: :string,
+                   desc: 'The name of the machine to operate on'
+      class_option :port,
+                   aliases: :p,
+                   type: :numeric,
+                   desc: 'The port that the machine is listening on for SSH connections'
+      class_option :limit,
+                   aliases: :l,
+                   type: :string,
+                   desc: 'Limits ansible runs to hosts matching PATTERN'
```
Comment (dacur) [inline lib/taperole/commands/ansible.rb]: you have this twice, i believe - see lines 7-10 above

File: bin/tape [hunk 2]
```diff
@@ -1,96 +1,5 @@
 #!/usr/bin/env ruby
 
-$LOAD_PATH << File.join(__dir__, '..', 'lib')
+require 'taperole'
```
Comment (dacur) [inline bin/tape]: how meta  💯

File: lib/taperole/commands/ansible.rb [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+      class_option :tags,
+                   aliases: :t,
+                   type: :string,
+                   desc: 'Only run plays and tasks tagged with these values'
```
Comment (dacur) [inline lib/taperole/commands/ansible.rb]: not sure if this is the same as `T` above

File: lib/taperole/commands/ansible.rb (hunk 1 above)
Comment (dkniffin) [inline lib/taperole/commands/ansible.rb]: Good catch. I think that was the case with the old code too. Maybe.

File: bin/tape (hunk 2 above)
Comment (dkniffin) [inline bin/tape]: :) That's the "right" :tm: way to do things. aka: I copied it from some other gems.

File: lib/taperole/commands/ansible.rb (hunk 3 above)
Comment (dkniffin) [inline lib/taperole/commands/ansible.rb]: Nah, it's different. `:t` allows you to only run certain ansible roles, whereas `:T` allows you to run a rake task on the server.

File: lib/taperole/core/ansible_runner.rb [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+      end
+    end
+
+    def valid_gems
+      has_gem_in_gemfile?('unicorn')
+    end
+
+    def has_gem_in_gemfile?(name)
+      if open('Gemfile').grep(/#{name}/).empty?
+        puts "💥 ERROR: Add #{name} to your Gemfile!💥 ".red
```
Comment (dacur) [inline lib/taperole/core/ansible_runner.rb]: 💯

Comment (dkniffin) [inline lib/taperole/core/ansible_runner.rb]: Oh, thanks for commenting on this. I missed a `logger.error` :)

**Comment by BrandonMathis** [comment]
@dkniffin THIS is pretty freakin D O P E

I like the overall structure of this code. Reads very well and make the running layer much easier to extend. Nice work 👍 

Only question I have is, does this change the structure of any of the cli commands?

**Comment by dkniffin** [comment]
@BrandonMathis I did remove some of the existing commands, because I don't think they were used. Specifically, [these](https://github.com/smashingboxes/taperole/blob/master/lib/tape/ansible_runner.rb#L6-L38). It looks like I also removed [the two that were added recently](https://github.com/smashingboxes/taperole/blob/master/lib/tape/ansible_runner.rb#L46-L51). Those should probably be added to the new code.

For `deploy` and `everything` I tried to keep everything as consistent with the current commands as possible.

**Comment by BrandonMathis** [comment]
Awesome. So freakin awesome. Thanks for doing this!

**Pull request merged**